### PR TITLE
Implement scans and reductions in pylibcudf

### PR DESCRIPTION
**Pull request by vyasr**
## Description

Contributes to #13921 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/_lib/pylibcudf/aggregation.pyx [hunk 1]
```diff
@@ -85,40 +83,43 @@ cdef class Aggregation:
...
 
-    cdef unique_ptr[groupby_aggregation] clone_underlying_as_groupby(self) except *:
-        """Make a copy of the underlying aggregation that can be used in a groupby.
+    cdef agg_ptr _raise_if_null(self, agg_ptr ptr, str alg):
+        # The functions calling this all use a dynamic cast between aggregation types,
+        # and the cast returning a null pointer is how we capture whether or not
+        # libcudf supports a given aggregation for a particular algorithm.
+        if ptr is NULL:
+            agg_repr = str(self.kind()).split(".")[1].title()
+            raise TypeError(f"{agg_repr} aggregations are not supported by {alg}")
```
Comment (shwina) [inline python/cudf/cudf/_lib/pylibcudf/aggregation.pyx:92]: I'd maybe reduce the scope of this helper:

```cython
cdef _unsupported_agg_error(self, alg):
    agg_repr = str(self.kind()).split(".")[1].title()
    raise TypeError(f"{agg_repr} aggregations are not supported by {alg}")   
```

Then:

```cython
cdef groupby_scan_aggregation *agg_cast = dynamic_cast[gbsa_ptr](agg.get())
if agg_cast is NULL:
    self._unsupported_agg_error("groupby scan")
```

This way we don't have to define the fused type `agg_ptr`

Comment (shwina) [reply python/cudf/cudf/_lib/pylibcudf/aggregation.pyx:92]: It's kinda subjective. Let me know if you think differently..

**Review by shwina — APPROVED** [review]
Generally looks great - one minor question/suggestion.

File: python/cudf/cudf/_lib/pylibcudf/aggregation.pyx (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/_lib/pylibcudf/aggregation.pyx:92]: Ha I actually had that originally and changed it to this to make the error consistent and reduce duplication, but then it adds the need for the fused type, which I agree feels heavy-handed here. Should we maybe just get rid of the helper entirely and just inline the error check?

Comment (vyasr) [reply python/cudf/cudf/_lib/pylibcudf/aggregation.pyx]: Discussed offline. Going with Ashwin's suggestion.

**Pull request merged**